mycpp/gc_str.cc

```diff
@@ -462,3 +468,108 @@ List<Str*>* Str::split(Str* sep) {
 
   return result;
 }
+
+static inline Str* _StrFormat(const char* fmt, int fmt_len, va_list args) {
+  auto beg = std::cregex_iterator(fmt, fmt + fmt_len, gStrFmtRegex);
+  auto end = std::cregex_iterator();
+
+  char int_buf[kMaxFmtWidth];
+  std::string buf;
+  for (std::cregex_iterator it = beg; it != end; ++it) {
+    const std::cmatch& match = *it;
+
+    const std::csub_match& lit_m = match[1];
+    assert(lit_m.matched);
+    const std::string& lit_s = lit_m.str();
+    buf.append(lit_s);
+
+    int width = 0;
+    bool zero_pad = false;
+    const std::csub_match& width_m = match[2];
+    const std::string& width_s = width_m.str();
+    if (width_m.matched && !width_s.empty()) {
+      if (width_s[0] == '0') {
+        zero_pad = true;
+        assert(width_s.size() > 1);
+        assert(StringToInteger(width_s.c_str() + 1, width_s.size() - 1, 10,
+                               &width));
+      } else {
+        assert(StringToInteger(width_s.c_str(), width_s.size(), 10, &width));
+      }
+      assert(width >= 0 && width < kMaxFmtWidth);
+    }
+
+    char const* str_to_add = nullptr;
+    int add_len = 0;
+    const std::csub_match& code_m = match[3];
+    const std::string& code_s = code_m.str();
+    if (!code_m.matched) {
+      assert(!width_m.matched);  // python errors on invalid format operators
+      goto done;
+    }
+    assert(code_s.size() == 1);
+    switch (code_s[0]) {
+    case '%': {
+      str_to_add = code_s.c_str();
+      add_len = 1;
+      break;
+    }
+    case 's': {
+      Str* s = va_arg(args, Str*);
+      str_to_add = s->data();
+      add_len = len(s);
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'r': {
+      Str* s = va_arg(args, Str*);
+      s = repr(s);
+      str_to_add = s->data();
+      add_len = len(s);
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'd':  // fallthrough
+    case 'o': {
+      int d = va_arg(args, int);
+      add_len = snprintf(int_buf, kMaxFmtWidth,
+                         match.str().c_str() + lit_s.size(), d);
+      assert(add_len > 0);
+      str_to_add = int_buf;
+      break;
+    }
+    default:
+      assert(0);
+      break;
+    }
+    assert(str_to_add != nullptr);
+
+  done:
+    if (str_to_add != nullptr) {
```

Comment:
I think if we change `goto done` to break, we don't need this nullptr check, because every other case will append something